### DISCUSSION TITLE
Using docker

### DISCUSSION THREAD

**Discussion by 19taurus79** [piccolo-orm/piccolo_admin#359] [Q&A]
Hi. Can you tell me how to use piccolo admin in docker. I have created a container, started admin application in the container, port forwarded, but I can't open it in browser. There was an idea to make an image for this purpose.

**Comment by dantownsend** [comment]
Do you get a 404 response in the browser?

Can you share some more info about your Dockerfile?

Also, if connecting to a database, make sure that your container can access the database, as that could cause 500 errors.

You're right, there's a plan to make a standalone Dockerised version of Piccolo Admin, but it's just a prototype at the moment and not generally available.

**Reply by 19taurus79** [reply]
Hi, thanks for the quick reply. I am getting "ERR_EMPTY_RESPONSE" from the browser. I have two containers, with an application where I use piccolo orm, and a database. I run through docker-compose and everything works. In the container with the application I have the admin file from your documentation example, locally I run it, the server comes up and everything works. When I run it in the container, via docker exec, I get a message that the server is running, but there is nothing on localhost:8000/admin.

**Reply by dantownsend** [reply]
Is port 8000 exposed in your Docker Compose file?

**Comment by 19taurus79** [comment]
rental_bot:
    container_name: rental_bot
    image: rental_bot_image
    build:
      context: bot
      dockerfile: Dockerfile
    env_file:
      - ./.env
    ports:
      - "8000:8000"
    depends_on:
      - db

**Reply by dantownsend** [reply]
All I can think of is:

- You might want to try running Piccolo Admin on `0.0.0.0` instead of localhost.
- Try the URL with / without a trailing slash, as this can sometimes be the problem.

**Reply by 19taurus79** [reply]
Thanks, I'll give it a try.

**Reply by 19taurus79** [reply]
Hi. Nothing's working ... It works locally, but it won't work through docker.

https://github.com/19taurus79/TestPiccoloInDocker.git

What am I doing wrong?

**Reply by 19taurus79** [reply]
I found the reason
main.py
if __name__ == "__main__":
    import uvicorn

    uvicorn.run("app:app", host="0.0.0.0",port=8000,reload=True)

**Reply by sinisaos** [reply]
@19taurus79 Sorry to bother you, from the last comment I'm not sure if you managed to solve the problem or not? [Here is a great article](https://testdriven.io/blog/fastapi-docker-traefik/) on how to use FastAPI and Docker and you can apply that to Piccolo Admin. I made small changes to `Dockerfile` and `docker-compose.yml` from your repo and everything works fine.
```dockerfile
# Dockerfile

FROM python:3.10-slim-bullseye

# Install the pip requirements:
RUN pip install --upgrade pip
ADD /requirements.txt /
RUN pip install -r /requirements.txt

# Add the application code:
COPY . .
```
```dockerfile
# docker-compose.yml

version: '3.9'

services:

  db:
    container_name: testing_db
    image: postgres:alpine3.19
    restart: always
    env_file:
     - .env
    volumes:
      - postgres_data:/var/lib/postgresql/data
    ports:
      - "5433:5432"

  test_app:
    container_name: test_app
    image: test_app_image
    build:
      context: back
      dockerfile: Dockerfile
    command: bash -c 'uvicorn app:app --host 0.0.0.0' # here you start app
    env_file:
      - .env
    ports:
      - "8001:8000"  
    depends_on:
      - db
volumes:
  postgres_data:
```

```bash
# .env

POSTGRES_PASSWORD=root
POSTGRES_USER=admin
POSTGRES_DB=test_db
```
Make migrations and create admin user with `docker exec -it your_test_app_docker_ps bash`
After that you can go to `http://localhost:8001/admin/` and use Piccolo Admin
Hope this helps.

**Reply by 19taurus79** [reply]
Thank you
